src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker.cs

```diff
@@ -0,0 +1,499 @@
+
+            private readonly Dictionary<ProjectId, ProjectFileInfo> _projectIdToFileInfoMap;
+            private readonly Dictionary<ProjectId, List<ProjectReference>> _projectIdToProjectReferencesMap;
+            private readonly Dictionary<string, ImmutableArray<ProjectInfo>> _pathToDiscoveredProjectInfosMap;
+
+            public Worker(
+                Workspace workspace,
+                DiagnosticReporter diagnosticReporter,
+                PathResolver pathResolver,
+                ProjectFileLoaderRegistry projectFileLoaderRegistry,
+                ProjectBuildManager buildManager,
+                ImmutableArray<string> requestedProjectPaths,
+                string baseDirectory,
+                ImmutableDictionary<string, string> globalProperties,
+                ProjectMap projectMap,
+                IProgress<ProjectLoadProgress> progress,
+                DiagnosticReportingOptions requestedProjectOptions,
+                DiagnosticReportingOptions discoveredProjectOptions,
+                bool preferMetadataForReferencedProjects)
+            {
+                _workspace = workspace;
+                _diagnosticReporter = diagnosticReporter;
+                _pathResolver = pathResolver;
+                _projectFileLoaderRegistry = projectFileLoaderRegistry;
+                _buildManager = buildManager;
+                _baseDirectory = baseDirectory;
+                _requestedProjectPaths = requestedProjectPaths;
+                _globalProperties = globalProperties;
+                _projectMap = projectMap ?? ProjectMap.Create();
+                _progress = progress;
+                _requestedProjectOptions = requestedProjectOptions;
+                _discoveredProjectOptions = discoveredProjectOptions;
+                _preferMetadataForReferencedProjects = preferMetadataForReferencedProjects;
+                _projectIdToFileInfoMap = new Dictionary<ProjectId, ProjectFileInfo>();
+                _pathToDiscoveredProjectInfosMap = new Dictionary<string, ImmutableArray<ProjectInfo>>(PathUtilities.Comparer);
+                _projectIdToProjectReferencesMap = new Dictionary<ProjectId, List<ProjectReference>>();
+            }
+
+            private async Task<TResult> DoOperationAndReportProgressAsync<TResult>(ProjectLoadOperation operation, string projectPath, string targetFramework, Func<Task<TResult>> doFunc)
+            {
+                var watch = _progress != null
+                    ? Stopwatch.StartNew()
+                    : null;
+
+                var result = await doFunc().ConfigureAwait(false);
+
+                if (_progress != null)
+                {
+                    watch.Stop();
+                    _progress.Report(new ProjectLoadProgress(projectPath, operation, targetFramework, watch.Elapsed));
+                }
+
+                return result;
+            }
+
+            public async Task<ImmutableArray<ProjectInfo>> LoadAsync(CancellationToken cancellationToken)
```

Comment:
> public async Task<ImmutableArray<ProjectInfo>> LoadAsync(CancellationToken cancellationToken) [](start = 11, length = 94)

Is it intended that the collection returned here has a non-determinstic order?

Reply:
Except in the MT case, there will be extra projects, right?  So the customer can't map from index in solution to index in result?

Reply:
Correct. That's always true.

Reply:
Consider the foreach over `_pathToDiscoveredProjectInfosMap`. Isn't that non-determinstic here as it's iterating a dictionary?

Reply:
Yes, that's what I said above:

> Following those will be any projects discovered while loading (e.g. referenced projects that aren't included in the solution), which will not be in a deterministic order.